File: src/main/java/com/hazza/LeetCodeByJava/No71SimplifyPath.java
package com.hazza.LeetCodeByJava;

import java.util.Deque;
import java.util.LinkedList;

/**
 * Created with IntelliJ IDEA.
 * Description: https://leetcode.com/problems/simplify-path/description/
 * Author: HazzaCheng
 * Contact: hazzacheng@gmail.com
 * Date: 2017-08-24
 * Time: 9:38 AM
 */
public class No71SimplifyPath {
    public String simplifyPath(String path) {
        String[] symbols = path.split("/+");
        Deque<String> queue = new LinkedList<>();
        for (String s: symbols) {
            if (s.equals("..") && !queue.isEmpty()) queue.pollLast();
            else if (!s.equals(".") && !s.equals("") && !s.equals("..")) queue.offerLast(s);
        }

        return "/" + String.join("/", queue);
    }
}

File: src/main/python/com/hazza/LeetCodeByPython/SegmentTwo/No71_Simplify_Path.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Created by HazzaCheng on 2019-07-02
# https://leetcode.com/problems/simplify-path/


class Solution:
    def simplifyPath(self, path: str) -> str:
        stack = []
        symbols = path.split('/')

        for s in symbols:
            if s == '..' and len(stack) != 0:
                stack.pop()
            elif s != '.' and s != '' and s != '..':
                stack.append(s)

        return '/' + '/'.join(stack)


if __name__ == '__main__':
    solution = Solution()

    print(solution.simplifyPath('/home/'))

    assert solution.simplifyPath('/home/') == '/home'
    assert solution.simplifyPath('/../') == '/'
    assert solution.simplifyPath('/...') == '/...'
    assert solution.simplifyPath('/home//foo/') == '/home/foo'
    assert solution.simplifyPath('/a/./b/../../c/') == '/c'
    assert solution.simplifyPath('/a/../../b/../c//.//') == '/c'
    assert solution.simplifyPath('/a//b////c/d//././/..') == '/a/b/c'
